### ISSUE TITLE
Allow a shared context to take window parameters independent of its parent

### ISSUE THREAD

**Issue by duncanburke** [tobbebex/GPipe-Core#19]
newSharedContext and runSharedContextT implicitly use the window configuration of the parent context. There should be a facility to pass in window configuration parameters independent of the parent configuration.

**Comment by plredmond** [comment]
I've already commented in general in https://github.com/tobbebex/GPipe-Core/issues/18, but since this more-specific thread is present, I'll propose a solution to start discussion.

@tobbebex, can we make both the context creation function *and* the shared-context creation function top-level functions in the implementing backend? (Currently the first is top-level, and the second exists in the `ContextHandle` data structure.)

Reasons to do this:

* (As far as I know) Only the application-specific code ever calls `newSharedContext`. GPipe doesn't need this to be in the `ContextHandle` structure.
* The backend implementation would be able to prefix *both* argument lists with backend-specific options, while still satisfying the `*Factory` type aliases defined in GPipe (as GPipe-GLFW [newContext'](https://github.com/plredmond/GPipe-GLFW/blob/master/GPipe-GLFW/src/Graphics/GPipe/Context/GLFW.hs#L73) already does).
* The existence of shared-context support would be optional for the implementing backend, allowing support for both advanced & simpler backends.
    * Simpler backends would have initially more-predictable behavior.
    * Simplifying the required API of a new context-creation backend would speed creation of backends, and thus adoption of GPipe.
* This would enable GPipe to support GLUT (which doesn't support shared contexts).

Reasons not to do this:

* We might not be able to do the safety-trick of the implicit object type which cannot escape, `os`.
* In general, the types might not work out- I haven't wrapped my head around everything GPipe does and don't know if it's possible.

Detail:

Change `Graphics.GPipe.Internal.Context` to something like:

```haskell
-- Context creation (spec)
type ContextFactory c ds w = ContextFormat c ds -> IO (ContextHandle w)

-- Shared-context creation (spec)
type SharedContextFactory c ds w os f m a = ContextHandle w -> ContextFormat c ds -> ContextT w os f m a

-- (Also remove `newSharedContext` from `ContextHandle`)
```

Change `Graphics.GPipe.Context.GLFW` to something like:

```haskell
-- Context creation (implementation): makes plain contexts
newContext :: ContextFactory c ds GLFWWindow
...
-- Context creation (implementation): makes contexts with backend-specific options
newContext' :: [WindowHint] -> WindowConf -> ContextFactory c ds GLFWWindow
...
-- Shared-context creation (implementation): makes shared-contexts with backend-specific options
newSharedContext' :: [WindowHint] -> WindowConf -> SharedContextFactory ... GLFWWindow ...
```

**Comment by tobbebex** [comment]
I have something similar implemented in c4bbb0e8129c382af4c7108f08f3ab5ba576854c in the ThreadedContexts branch, but using a specific SharedContextFactory type is indeed better for the GLUT case!

Either way, this will require a major version bump, so before I do this I'd like to finish #13 that can be made with only a minor bump.

**Comment by tobbebex** [comment]
Closing this as implemented with 32ef0ec6b1208ec53ab163976e40bf9b54c41ab5 and to be released in 2.2.
